# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Functionta
    end_sequence_id = target_vocab_to_int['<EOS>']
    
    source_id_text = [[source_vocab_to_int[vocab] for vocab in line.split()] for line in 
                      source_text.split('\n')]
    
    target_id_text =  [[target_vocab_to_int[vocab] for vocab in line.split()] + [end_sequence_id] 
                       for line in target_text.split('\n')]
    
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.2.1
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    inputs = tf.placeholder(tf.int32, shape=[None, None], name='input')
    targets = tf.placeholder(tf.int32, shape=[None, None], name='targets')
    learning_rate = tf.placeholder(tf.float32, name='learning_rate')
    keep_prob = tf.placeholder(tf.float32, name = 'keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, shape=[None], name = 'target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name = 'max_target_len')
    source_sequence_length = tf.placeholder(tf.int32,  shape=[None], name = 'source_sequence_length')
    return inputs, targets, learning_rate, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Operation'>):
<tf.Operation 'assert_rank_2/Assert/Assert' type=Assert>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "/usr/local/lib/python3.5/runpy.py", line 193, in _run_module_as_main\n    "__main__", mod_spec)', 'File "/usr/local/lib/python3.5/runpy.py", line 85, in _run_code\n    exec(code, run_globals)', 'File "/usr/local/lib/python3.5/site-packages/ipykernel_launcher.py", line 16, in <module>\n    app.launch_new_instance()', 'File "/usr/local/lib/python3.5/site-packages/traitlets/config/application.py", line 658, in launch_instance\n    app.start()', 'File "/usr/local/lib/python3.5/site-packages/ipykernel/kernelapp.py", line 477, in start\n    ioloop.IOLoop.instance().start()', 'File "/usr/local/lib/python3.5/site-packages/zmq/eventloop/ioloop.py", line 177, in start\n    super(ZMQIOLoop, self).start()', 'F

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    start_sequence_id = target_vocab_to_int['<GO>']
    
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    decoding_input = tf.concat([tf.fill([batch_size, 1], start_sequence_id), ending], 1)

    return decoding_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    def build_cell(rnn_size, keep_prob):
        lstm = tf.contrib.rnn.LSTMCell(rnn_size)
        lstm_drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
        return lstm_drop
    # Stack them all
    stacked_lstm = tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    # Creathe embedding layer.
    embed_encoder = tf.contrib.layers.embed_sequence(rnn_inputs, vocab_size = source_vocab_size, embed_dim = encoding_embedding_size)
    # If we don't have an initial zero state, provide a dtype.
    output, state = tf.nn.dynamic_rnn(stacked_lstm, embed_encoder, source_sequence_length, dtype=tf.float32)
    return output, state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:
def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    trainig_helper = tf.contrib.seq2seq.TrainingHelper(dec_embed_input, target_sequence_length)
    basic_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, trainig_helper, encoder_state, output_layer)
    f_output, _, _ = tf.contrib.seq2seq.dynamic_decode(basic_decoder,maximum_iterations=max_summary_length)
    return f_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    start_ids = tf.tile([start_of_sequence_id], [batch_size])
    # Create the embedding helper.
    embedding_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(
        dec_embeddings, start_ids, end_of_sequence_id)
    basic_decoder = tf.contrib.seq2seq.BasicDecoder(
        dec_cell, embedding_helper, encoder_state, output_layer)
    f_output, _, _ = tf.contrib.seq2seq.dynamic_decode(
        basic_decoder,maximum_iterations=max_target_sequence_length)
    return f_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # Use the same proess as in the encoding layer.
    def build_cell(rnn_size, keep_prob):
        lstm = tf.contrib.rnn.LSTMCell(rnn_size)
        lstm_drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
        return lstm_drop
    # Stack them all
    stacked_lstm = tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

    dense_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    
    with tf.variable_scope("decode") as scope:
        tr_decoder_output = decoding_layer_train(
            encoder_state, stacked_lstm, dec_embed_input, 
            target_sequence_length, max_target_sequence_length, 
            dense_layer, keep_prob)
        scope.reuse_variables()
        inf_decoder_output = decoding_layer_infer(
            encoder_state, stacked_lstm, dec_embeddings, 
            target_vocab_to_int['<GO>'], target_vocab_to_int['<EOS>'], 
            max_target_sequence_length, target_vocab_size, 
            dense_layer, batch_size, keep_prob)
    
    return tr_decoder_output, inf_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    output, state = encoding_layer(input_data, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   enc_embedding_size)
    
    processed_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    tr_decoder_output, inf_decoder_output = decoding_layer(processed_input, state,
                   target_sequence_length, max_target_sentence_length,
                   rnn_size, num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, dec_embedding_size)
    
    return tr_decoder_output, inf_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [14]:
# Number of Epochs
epochs = 10
# Batch Size
batch_size = 512
# RNN Size
rnn_size = 128
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 128
decoding_embedding_size = 128
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.55
display_step = True

### Build the Graph
Build the graph using the neural network you implemented.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    1/269 - Train Accuracy: 0.2329, Validation Accuracy: 0.3096, Loss: 5.6685
Epoch   0 Batch    2/269 - Train Accuracy: 0.2655, Validation Accuracy: 0.3096, Loss: 5.4388
Epoch   0 Batch    3/269 - Train Accuracy: 0.2444, Validation Accuracy: 0.3096, Loss: 5.2835
Epoch   0 Batch    4/269 - Train Accuracy: 0.2317, Validation Accuracy: 0.3096, Loss: 5.1198
Epoch   0 Batch    5/269 - Train Accuracy: 0.2324, Validation Accuracy: 0.3096, Loss: 4.9409
Epoch   0 Batch    6/269 - Train Accuracy: 0.2785, Validation Accuracy: 0.3096, Loss: 4.6206
Epoch   0 Batch    7/269 - Train Accuracy: 0.2765, Validation Accuracy: 0.3096, Loss: 4.4923
Epoch   0 Batch    8/269 - Train Accuracy: 0.2422, Validation Accuracy: 0.3106, Loss: 4.5626
Epoch   0 Batch    9/269 - Train Accuracy: 0.2797, Validation Accuracy: 0.3220, Loss: 4.3633
Epoch   0 Batch   10/269 - Train Accuracy: 0.2617, Validation Accuracy: 0.3357, Loss: 4.4186
Epoch   0 Batch   11/269 - Train Accuracy: 0.3048, Validation Accuracy

Epoch   0 Batch   90/269 - Train Accuracy: 0.4353, Validation Accuracy: 0.4906, Loss: 2.5805
Epoch   0 Batch   91/269 - Train Accuracy: 0.4405, Validation Accuracy: 0.4720, Loss: 2.4340
Epoch   0 Batch   92/269 - Train Accuracy: 0.4621, Validation Accuracy: 0.4869, Loss: 2.4352
Epoch   0 Batch   93/269 - Train Accuracy: 0.4827, Validation Accuracy: 0.4832, Loss: 2.3316
Epoch   0 Batch   94/269 - Train Accuracy: 0.4501, Validation Accuracy: 0.4798, Loss: 2.4261
Epoch   0 Batch   95/269 - Train Accuracy: 0.4631, Validation Accuracy: 0.4869, Loss: 2.4020
Epoch   0 Batch   96/269 - Train Accuracy: 0.4650, Validation Accuracy: 0.4900, Loss: 2.3894
Epoch   0 Batch   97/269 - Train Accuracy: 0.4623, Validation Accuracy: 0.4888, Loss: 2.3872
Epoch   0 Batch   98/269 - Train Accuracy: 0.4674, Validation Accuracy: 0.4830, Loss: 2.3332
Epoch   0 Batch   99/269 - Train Accuracy: 0.4311, Validation Accuracy: 0.4810, Loss: 2.5105
Epoch   0 Batch  100/269 - Train Accuracy: 0.4821, Validation Accuracy

Epoch   0 Batch  179/269 - Train Accuracy: 0.4856, Validation Accuracy: 0.5065, Loss: 1.7604
Epoch   0 Batch  180/269 - Train Accuracy: 0.4905, Validation Accuracy: 0.5130, Loss: 1.7526
Epoch   0 Batch  181/269 - Train Accuracy: 0.4794, Validation Accuracy: 0.5074, Loss: 1.7372
Epoch   0 Batch  182/269 - Train Accuracy: 0.4819, Validation Accuracy: 0.5040, Loss: 1.7635
Epoch   0 Batch  183/269 - Train Accuracy: 0.5540, Validation Accuracy: 0.5105, Loss: 1.5173
Epoch   0 Batch  184/269 - Train Accuracy: 0.4503, Validation Accuracy: 0.4971, Loss: 1.8036
Epoch   0 Batch  185/269 - Train Accuracy: 0.4962, Validation Accuracy: 0.5144, Loss: 1.7125
Epoch   0 Batch  186/269 - Train Accuracy: 0.4600, Validation Accuracy: 0.5131, Loss: 1.8033
Epoch   0 Batch  187/269 - Train Accuracy: 0.4802, Validation Accuracy: 0.5036, Loss: 1.6867
Epoch   0 Batch  188/269 - Train Accuracy: 0.5041, Validation Accuracy: 0.5123, Loss: 1.6626
Epoch   0 Batch  189/269 - Train Accuracy: 0.4976, Validation Accuracy

Epoch   1 Batch    1/269 - Train Accuracy: 0.4520, Validation Accuracy: 0.5034, Loss: 1.4224
Epoch   1 Batch    2/269 - Train Accuracy: 0.4828, Validation Accuracy: 0.5170, Loss: 1.3835
Epoch   1 Batch    3/269 - Train Accuracy: 0.4771, Validation Accuracy: 0.5191, Loss: 1.4114
Epoch   1 Batch    4/269 - Train Accuracy: 0.4548, Validation Accuracy: 0.5115, Loss: 1.4039
Epoch   1 Batch    5/269 - Train Accuracy: 0.4732, Validation Accuracy: 0.5224, Loss: 1.4193
Epoch   1 Batch    6/269 - Train Accuracy: 0.4993, Validation Accuracy: 0.5130, Loss: 1.2883
Epoch   1 Batch    7/269 - Train Accuracy: 0.4898, Validation Accuracy: 0.5044, Loss: 1.3332
Epoch   1 Batch    8/269 - Train Accuracy: 0.4832, Validation Accuracy: 0.5162, Loss: 1.3944
Epoch   1 Batch    9/269 - Train Accuracy: 0.5026, Validation Accuracy: 0.5297, Loss: 1.3299
Epoch   1 Batch   10/269 - Train Accuracy: 0.4631, Validation Accuracy: 0.5083, Loss: 1.3646
Epoch   1 Batch   11/269 - Train Accuracy: 0.4865, Validation Accuracy

Epoch   1 Batch   90/269 - Train Accuracy: 0.4991, Validation Accuracy: 0.5503, Loss: 1.1408
Epoch   1 Batch   91/269 - Train Accuracy: 0.5356, Validation Accuracy: 0.5580, Loss: 1.0617
Epoch   1 Batch   92/269 - Train Accuracy: 0.5498, Validation Accuracy: 0.5653, Loss: 1.0632
Epoch   1 Batch   93/269 - Train Accuracy: 0.5576, Validation Accuracy: 0.5621, Loss: 1.0320
Epoch   1 Batch   94/269 - Train Accuracy: 0.5384, Validation Accuracy: 0.5588, Loss: 1.0844
Epoch   1 Batch   95/269 - Train Accuracy: 0.5429, Validation Accuracy: 0.5534, Loss: 1.0685
Epoch   1 Batch   96/269 - Train Accuracy: 0.5418, Validation Accuracy: 0.5541, Loss: 1.0529
Epoch   1 Batch   97/269 - Train Accuracy: 0.5283, Validation Accuracy: 0.5531, Loss: 1.0634
Epoch   1 Batch   98/269 - Train Accuracy: 0.5563, Validation Accuracy: 0.5468, Loss: 1.0343
Epoch   1 Batch   99/269 - Train Accuracy: 0.5178, Validation Accuracy: 0.5464, Loss: 1.1038
Epoch   1 Batch  100/269 - Train Accuracy: 0.5506, Validation Accuracy

Epoch   1 Batch  178/269 - Train Accuracy: 0.5464, Validation Accuracy: 0.5718, Loss: 0.9380
Epoch   1 Batch  179/269 - Train Accuracy: 0.5645, Validation Accuracy: 0.5748, Loss: 0.9147
Epoch   1 Batch  180/269 - Train Accuracy: 0.5596, Validation Accuracy: 0.5757, Loss: 0.9009
Epoch   1 Batch  181/269 - Train Accuracy: 0.5555, Validation Accuracy: 0.5823, Loss: 0.9092
Epoch   1 Batch  182/269 - Train Accuracy: 0.5733, Validation Accuracy: 0.5844, Loss: 0.9161
Epoch   1 Batch  183/269 - Train Accuracy: 0.6323, Validation Accuracy: 0.5803, Loss: 0.7896
Epoch   1 Batch  184/269 - Train Accuracy: 0.5486, Validation Accuracy: 0.5785, Loss: 0.9406
Epoch   1 Batch  185/269 - Train Accuracy: 0.5717, Validation Accuracy: 0.5776, Loss: 0.9010
Epoch   1 Batch  186/269 - Train Accuracy: 0.5408, Validation Accuracy: 0.5811, Loss: 0.9319
Epoch   1 Batch  187/269 - Train Accuracy: 0.5620, Validation Accuracy: 0.5748, Loss: 0.8895
Epoch   1 Batch  188/269 - Train Accuracy: 0.5702, Validation Accuracy

Epoch   1 Batch  267/269 - Train Accuracy: 0.5951, Validation Accuracy: 0.6007, Loss: 0.8264
Epoch   2 Batch    1/269 - Train Accuracy: 0.5821, Validation Accuracy: 0.6056, Loss: 0.8273
Epoch   2 Batch    2/269 - Train Accuracy: 0.5809, Validation Accuracy: 0.6049, Loss: 0.8189
Epoch   2 Batch    3/269 - Train Accuracy: 0.5971, Validation Accuracy: 0.6037, Loss: 0.8227
Epoch   2 Batch    4/269 - Train Accuracy: 0.5663, Validation Accuracy: 0.6040, Loss: 0.8301
Epoch   2 Batch    5/269 - Train Accuracy: 0.5772, Validation Accuracy: 0.6040, Loss: 0.8358
Epoch   2 Batch    6/269 - Train Accuracy: 0.6027, Validation Accuracy: 0.6053, Loss: 0.7681
Epoch   2 Batch    7/269 - Train Accuracy: 0.6032, Validation Accuracy: 0.6042, Loss: 0.7884
Epoch   2 Batch    8/269 - Train Accuracy: 0.5812, Validation Accuracy: 0.6049, Loss: 0.8356
Epoch   2 Batch    9/269 - Train Accuracy: 0.5924, Validation Accuracy: 0.6026, Loss: 0.8106
Epoch   2 Batch   10/269 - Train Accuracy: 0.5882, Validation Accuracy

Epoch   2 Batch   89/269 - Train Accuracy: 0.6341, Validation Accuracy: 0.6266, Loss: 0.7269
Epoch   2 Batch   90/269 - Train Accuracy: 0.5802, Validation Accuracy: 0.6271, Loss: 0.7715
Epoch   2 Batch   91/269 - Train Accuracy: 0.6250, Validation Accuracy: 0.6246, Loss: 0.7163
Epoch   2 Batch   92/269 - Train Accuracy: 0.6175, Validation Accuracy: 0.6226, Loss: 0.7206
Epoch   2 Batch   93/269 - Train Accuracy: 0.6305, Validation Accuracy: 0.6308, Loss: 0.6966
Epoch   2 Batch   94/269 - Train Accuracy: 0.6201, Validation Accuracy: 0.6277, Loss: 0.7417
Epoch   2 Batch   95/269 - Train Accuracy: 0.6070, Validation Accuracy: 0.6222, Loss: 0.7267
Epoch   2 Batch   96/269 - Train Accuracy: 0.6179, Validation Accuracy: 0.6229, Loss: 0.7283
Epoch   2 Batch   97/269 - Train Accuracy: 0.6070, Validation Accuracy: 0.6246, Loss: 0.7211
Epoch   2 Batch   98/269 - Train Accuracy: 0.6253, Validation Accuracy: 0.6247, Loss: 0.7188
Epoch   2 Batch   99/269 - Train Accuracy: 0.6106, Validation Accuracy

Epoch   2 Batch  178/269 - Train Accuracy: 0.6131, Validation Accuracy: 0.6333, Loss: 0.6875
Epoch   2 Batch  179/269 - Train Accuracy: 0.6296, Validation Accuracy: 0.6326, Loss: 0.6752
Epoch   2 Batch  180/269 - Train Accuracy: 0.6247, Validation Accuracy: 0.6318, Loss: 0.6612
Epoch   2 Batch  181/269 - Train Accuracy: 0.6164, Validation Accuracy: 0.6335, Loss: 0.6765
Epoch   2 Batch  182/269 - Train Accuracy: 0.6301, Validation Accuracy: 0.6309, Loss: 0.6665
Epoch   2 Batch  183/269 - Train Accuracy: 0.6713, Validation Accuracy: 0.6341, Loss: 0.5858
Epoch   2 Batch  184/269 - Train Accuracy: 0.5997, Validation Accuracy: 0.6326, Loss: 0.6960
Epoch   2 Batch  185/269 - Train Accuracy: 0.6333, Validation Accuracy: 0.6312, Loss: 0.6681
Epoch   2 Batch  186/269 - Train Accuracy: 0.6056, Validation Accuracy: 0.6305, Loss: 0.6846
Epoch   2 Batch  187/269 - Train Accuracy: 0.6344, Validation Accuracy: 0.6377, Loss: 0.6516
Epoch   2 Batch  188/269 - Train Accuracy: 0.6311, Validation Accuracy

Epoch   2 Batch  267/269 - Train Accuracy: 0.6401, Validation Accuracy: 0.6308, Loss: 0.6239
Epoch   3 Batch    1/269 - Train Accuracy: 0.6127, Validation Accuracy: 0.6231, Loss: 0.6338
Epoch   3 Batch    2/269 - Train Accuracy: 0.6012, Validation Accuracy: 0.6244, Loss: 0.6195
Epoch   3 Batch    3/269 - Train Accuracy: 0.6314, Validation Accuracy: 0.6294, Loss: 0.6219
Epoch   3 Batch    4/269 - Train Accuracy: 0.6045, Validation Accuracy: 0.6380, Loss: 0.6361
Epoch   3 Batch    5/269 - Train Accuracy: 0.6158, Validation Accuracy: 0.6253, Loss: 0.6379
Epoch   3 Batch    6/269 - Train Accuracy: 0.6305, Validation Accuracy: 0.6276, Loss: 0.5898
Epoch   3 Batch    7/269 - Train Accuracy: 0.6306, Validation Accuracy: 0.6275, Loss: 0.5964
Epoch   3 Batch    8/269 - Train Accuracy: 0.6025, Validation Accuracy: 0.6286, Loss: 0.6398
Epoch   3 Batch    9/269 - Train Accuracy: 0.6353, Validation Accuracy: 0.6392, Loss: 0.6260
Epoch   3 Batch   10/269 - Train Accuracy: 0.6152, Validation Accuracy

Epoch   3 Batch   89/269 - Train Accuracy: 0.6621, Validation Accuracy: 0.6475, Loss: 0.5654
Epoch   3 Batch   90/269 - Train Accuracy: 0.6354, Validation Accuracy: 0.6640, Loss: 0.6014
Epoch   3 Batch   91/269 - Train Accuracy: 0.6705, Validation Accuracy: 0.6644, Loss: 0.5476
Epoch   3 Batch   92/269 - Train Accuracy: 0.6465, Validation Accuracy: 0.6511, Loss: 0.5620
Epoch   3 Batch   93/269 - Train Accuracy: 0.6546, Validation Accuracy: 0.6502, Loss: 0.5423
Epoch   3 Batch   94/269 - Train Accuracy: 0.6465, Validation Accuracy: 0.6480, Loss: 0.5776
Epoch   3 Batch   95/269 - Train Accuracy: 0.6523, Validation Accuracy: 0.6591, Loss: 0.5602
Epoch   3 Batch   96/269 - Train Accuracy: 0.6655, Validation Accuracy: 0.6586, Loss: 0.5600
Epoch   3 Batch   97/269 - Train Accuracy: 0.6590, Validation Accuracy: 0.6538, Loss: 0.5537
Epoch   3 Batch   98/269 - Train Accuracy: 0.6504, Validation Accuracy: 0.6378, Loss: 0.5580
Epoch   3 Batch   99/269 - Train Accuracy: 0.6265, Validation Accuracy

Epoch   3 Batch  178/269 - Train Accuracy: 0.6808, Validation Accuracy: 0.6867, Loss: 0.5392
Epoch   3 Batch  179/269 - Train Accuracy: 0.6740, Validation Accuracy: 0.6714, Loss: 0.5315
Epoch   3 Batch  180/269 - Train Accuracy: 0.6785, Validation Accuracy: 0.6721, Loss: 0.5168
Epoch   3 Batch  181/269 - Train Accuracy: 0.6833, Validation Accuracy: 0.6935, Loss: 0.5270
Epoch   3 Batch  182/269 - Train Accuracy: 0.6807, Validation Accuracy: 0.6726, Loss: 0.5248
Epoch   3 Batch  183/269 - Train Accuracy: 0.7155, Validation Accuracy: 0.6639, Loss: 0.4553
Epoch   3 Batch  184/269 - Train Accuracy: 0.6760, Validation Accuracy: 0.6881, Loss: 0.5464
Epoch   3 Batch  185/269 - Train Accuracy: 0.6871, Validation Accuracy: 0.6823, Loss: 0.5133
Epoch   3 Batch  186/269 - Train Accuracy: 0.6535, Validation Accuracy: 0.6728, Loss: 0.5340
Epoch   3 Batch  187/269 - Train Accuracy: 0.6811, Validation Accuracy: 0.6712, Loss: 0.5075
Epoch   3 Batch  188/269 - Train Accuracy: 0.6850, Validation Accuracy

Epoch   3 Batch  267/269 - Train Accuracy: 0.6990, Validation Accuracy: 0.7068, Loss: 0.4998
Epoch   4 Batch    1/269 - Train Accuracy: 0.7016, Validation Accuracy: 0.7025, Loss: 0.4960
Epoch   4 Batch    2/269 - Train Accuracy: 0.6892, Validation Accuracy: 0.7002, Loss: 0.4842
Epoch   4 Batch    3/269 - Train Accuracy: 0.7120, Validation Accuracy: 0.7069, Loss: 0.4961
Epoch   4 Batch    4/269 - Train Accuracy: 0.6848, Validation Accuracy: 0.7087, Loss: 0.5102
Epoch   4 Batch    5/269 - Train Accuracy: 0.6703, Validation Accuracy: 0.7029, Loss: 0.5044
Epoch   4 Batch    6/269 - Train Accuracy: 0.7153, Validation Accuracy: 0.7031, Loss: 0.4709
Epoch   4 Batch    7/269 - Train Accuracy: 0.7074, Validation Accuracy: 0.7072, Loss: 0.4731
Epoch   4 Batch    8/269 - Train Accuracy: 0.6931, Validation Accuracy: 0.7014, Loss: 0.5114
Epoch   4 Batch    9/269 - Train Accuracy: 0.7260, Validation Accuracy: 0.7146, Loss: 0.5024
Epoch   4 Batch   10/269 - Train Accuracy: 0.7103, Validation Accuracy

Epoch   4 Batch   89/269 - Train Accuracy: 0.7355, Validation Accuracy: 0.7353, Loss: 0.4516
Epoch   4 Batch   90/269 - Train Accuracy: 0.6997, Validation Accuracy: 0.7271, Loss: 0.4718
Epoch   4 Batch   91/269 - Train Accuracy: 0.7191, Validation Accuracy: 0.7081, Loss: 0.4449
Epoch   4 Batch   92/269 - Train Accuracy: 0.7053, Validation Accuracy: 0.7082, Loss: 0.4404
Epoch   4 Batch   93/269 - Train Accuracy: 0.7341, Validation Accuracy: 0.7256, Loss: 0.4302
Epoch   4 Batch   94/269 - Train Accuracy: 0.7156, Validation Accuracy: 0.7282, Loss: 0.4544
Epoch   4 Batch   95/269 - Train Accuracy: 0.7062, Validation Accuracy: 0.7124, Loss: 0.4423
Epoch   4 Batch   96/269 - Train Accuracy: 0.7082, Validation Accuracy: 0.7164, Loss: 0.4461
Epoch   4 Batch   97/269 - Train Accuracy: 0.7354, Validation Accuracy: 0.7298, Loss: 0.4355
Epoch   4 Batch   98/269 - Train Accuracy: 0.7437, Validation Accuracy: 0.7322, Loss: 0.4489
Epoch   4 Batch   99/269 - Train Accuracy: 0.7092, Validation Accuracy

Epoch   4 Batch  178/269 - Train Accuracy: 0.7530, Validation Accuracy: 0.7479, Loss: 0.4172
Epoch   4 Batch  179/269 - Train Accuracy: 0.7272, Validation Accuracy: 0.7436, Loss: 0.4131
Epoch   4 Batch  180/269 - Train Accuracy: 0.7432, Validation Accuracy: 0.7258, Loss: 0.3998
Epoch   4 Batch  181/269 - Train Accuracy: 0.7441, Validation Accuracy: 0.7517, Loss: 0.4210
Epoch   4 Batch  182/269 - Train Accuracy: 0.7540, Validation Accuracy: 0.7468, Loss: 0.4106
Epoch   4 Batch  183/269 - Train Accuracy: 0.7887, Validation Accuracy: 0.7412, Loss: 0.3568
Epoch   4 Batch  184/269 - Train Accuracy: 0.7335, Validation Accuracy: 0.7351, Loss: 0.4186
Epoch   4 Batch  185/269 - Train Accuracy: 0.7605, Validation Accuracy: 0.7457, Loss: 0.3993
Epoch   4 Batch  186/269 - Train Accuracy: 0.7501, Validation Accuracy: 0.7472, Loss: 0.4138
Epoch   4 Batch  187/269 - Train Accuracy: 0.7477, Validation Accuracy: 0.7508, Loss: 0.3947
Epoch   4 Batch  188/269 - Train Accuracy: 0.7428, Validation Accuracy

Epoch   4 Batch  267/269 - Train Accuracy: 0.7664, Validation Accuracy: 0.7706, Loss: 0.3929
Epoch   5 Batch    1/269 - Train Accuracy: 0.7840, Validation Accuracy: 0.7616, Loss: 0.3916
Epoch   5 Batch    2/269 - Train Accuracy: 0.7537, Validation Accuracy: 0.7538, Loss: 0.3920
Epoch   5 Batch    3/269 - Train Accuracy: 0.7716, Validation Accuracy: 0.7632, Loss: 0.3875
Epoch   5 Batch    4/269 - Train Accuracy: 0.7569, Validation Accuracy: 0.7591, Loss: 0.3920
Epoch   5 Batch    5/269 - Train Accuracy: 0.7487, Validation Accuracy: 0.7604, Loss: 0.3928
Epoch   5 Batch    6/269 - Train Accuracy: 0.7472, Validation Accuracy: 0.7514, Loss: 0.3653
Epoch   5 Batch    7/269 - Train Accuracy: 0.7758, Validation Accuracy: 0.7758, Loss: 0.3677
Epoch   5 Batch    8/269 - Train Accuracy: 0.7719, Validation Accuracy: 0.7701, Loss: 0.3984
Epoch   5 Batch    9/269 - Train Accuracy: 0.7743, Validation Accuracy: 0.7657, Loss: 0.3854
Epoch   5 Batch   10/269 - Train Accuracy: 0.7562, Validation Accuracy

Epoch   5 Batch   89/269 - Train Accuracy: 0.8105, Validation Accuracy: 0.7971, Loss: 0.3430
Epoch   5 Batch   90/269 - Train Accuracy: 0.7739, Validation Accuracy: 0.7924, Loss: 0.3577
Epoch   5 Batch   91/269 - Train Accuracy: 0.8074, Validation Accuracy: 0.7975, Loss: 0.3367
Epoch   5 Batch   92/269 - Train Accuracy: 0.8073, Validation Accuracy: 0.7928, Loss: 0.3351
Epoch   5 Batch   93/269 - Train Accuracy: 0.8034, Validation Accuracy: 0.8018, Loss: 0.3359
Epoch   5 Batch   94/269 - Train Accuracy: 0.7868, Validation Accuracy: 0.7946, Loss: 0.3566
Epoch   5 Batch   95/269 - Train Accuracy: 0.8016, Validation Accuracy: 0.8095, Loss: 0.3439
Epoch   5 Batch   96/269 - Train Accuracy: 0.7877, Validation Accuracy: 0.7967, Loss: 0.3406
Epoch   5 Batch   97/269 - Train Accuracy: 0.8079, Validation Accuracy: 0.8047, Loss: 0.3405
Epoch   5 Batch   98/269 - Train Accuracy: 0.7998, Validation Accuracy: 0.7974, Loss: 0.3411
Epoch   5 Batch   99/269 - Train Accuracy: 0.7900, Validation Accuracy

Epoch   5 Batch  178/269 - Train Accuracy: 0.8253, Validation Accuracy: 0.8179, Loss: 0.3232
Epoch   5 Batch  179/269 - Train Accuracy: 0.8171, Validation Accuracy: 0.8211, Loss: 0.3206
Epoch   5 Batch  180/269 - Train Accuracy: 0.8180, Validation Accuracy: 0.8167, Loss: 0.3113
Epoch   5 Batch  181/269 - Train Accuracy: 0.7991, Validation Accuracy: 0.8169, Loss: 0.3291
Epoch   5 Batch  182/269 - Train Accuracy: 0.8250, Validation Accuracy: 0.8249, Loss: 0.3196
Epoch   5 Batch  183/269 - Train Accuracy: 0.8436, Validation Accuracy: 0.8214, Loss: 0.2782
Epoch   5 Batch  184/269 - Train Accuracy: 0.8122, Validation Accuracy: 0.8185, Loss: 0.3250
Epoch   5 Batch  185/269 - Train Accuracy: 0.8334, Validation Accuracy: 0.8196, Loss: 0.3095
Epoch   5 Batch  186/269 - Train Accuracy: 0.8114, Validation Accuracy: 0.8226, Loss: 0.3244
Epoch   5 Batch  187/269 - Train Accuracy: 0.8140, Validation Accuracy: 0.8202, Loss: 0.3121
Epoch   5 Batch  188/269 - Train Accuracy: 0.8367, Validation Accuracy

Epoch   5 Batch  267/269 - Train Accuracy: 0.8279, Validation Accuracy: 0.8374, Loss: 0.3108
Epoch   6 Batch    1/269 - Train Accuracy: 0.8438, Validation Accuracy: 0.8399, Loss: 0.2987
Epoch   6 Batch    2/269 - Train Accuracy: 0.8288, Validation Accuracy: 0.8343, Loss: 0.3018
Epoch   6 Batch    3/269 - Train Accuracy: 0.8352, Validation Accuracy: 0.8344, Loss: 0.2997
Epoch   6 Batch    4/269 - Train Accuracy: 0.8232, Validation Accuracy: 0.8319, Loss: 0.3089
Epoch   6 Batch    5/269 - Train Accuracy: 0.8307, Validation Accuracy: 0.8368, Loss: 0.2997
Epoch   6 Batch    6/269 - Train Accuracy: 0.8621, Validation Accuracy: 0.8368, Loss: 0.2790
Epoch   6 Batch    7/269 - Train Accuracy: 0.8422, Validation Accuracy: 0.8422, Loss: 0.2827
Epoch   6 Batch    8/269 - Train Accuracy: 0.8425, Validation Accuracy: 0.8429, Loss: 0.3004
Epoch   6 Batch    9/269 - Train Accuracy: 0.8285, Validation Accuracy: 0.8408, Loss: 0.3037
Epoch   6 Batch   10/269 - Train Accuracy: 0.8296, Validation Accuracy

Epoch   6 Batch   89/269 - Train Accuracy: 0.8662, Validation Accuracy: 0.8485, Loss: 0.2758
Epoch   6 Batch   90/269 - Train Accuracy: 0.8328, Validation Accuracy: 0.8426, Loss: 0.2870
Epoch   6 Batch   91/269 - Train Accuracy: 0.8643, Validation Accuracy: 0.8445, Loss: 0.2689
Epoch   6 Batch   92/269 - Train Accuracy: 0.8675, Validation Accuracy: 0.8469, Loss: 0.2622
Epoch   6 Batch   93/269 - Train Accuracy: 0.8478, Validation Accuracy: 0.8432, Loss: 0.2585
Epoch   6 Batch   94/269 - Train Accuracy: 0.8341, Validation Accuracy: 0.8451, Loss: 0.2894
Epoch   6 Batch   95/269 - Train Accuracy: 0.8456, Validation Accuracy: 0.8493, Loss: 0.2671
Epoch   6 Batch   96/269 - Train Accuracy: 0.8352, Validation Accuracy: 0.8445, Loss: 0.2757
Epoch   6 Batch   97/269 - Train Accuracy: 0.8591, Validation Accuracy: 0.8442, Loss: 0.2658
Epoch   6 Batch   98/269 - Train Accuracy: 0.8503, Validation Accuracy: 0.8443, Loss: 0.2634
Epoch   6 Batch   99/269 - Train Accuracy: 0.8375, Validation Accuracy

Epoch   6 Batch  178/269 - Train Accuracy: 0.8577, Validation Accuracy: 0.8557, Loss: 0.2582
Epoch   6 Batch  179/269 - Train Accuracy: 0.8517, Validation Accuracy: 0.8549, Loss: 0.2517
Epoch   6 Batch  180/269 - Train Accuracy: 0.8631, Validation Accuracy: 0.8501, Loss: 0.2456
Epoch   6 Batch  181/269 - Train Accuracy: 0.8459, Validation Accuracy: 0.8483, Loss: 0.2586
Epoch   6 Batch  182/269 - Train Accuracy: 0.8584, Validation Accuracy: 0.8547, Loss: 0.2533
Epoch   6 Batch  183/269 - Train Accuracy: 0.8767, Validation Accuracy: 0.8563, Loss: 0.2269
Epoch   6 Batch  184/269 - Train Accuracy: 0.8622, Validation Accuracy: 0.8509, Loss: 0.2620
Epoch   6 Batch  185/269 - Train Accuracy: 0.8685, Validation Accuracy: 0.8561, Loss: 0.2493
Epoch   6 Batch  186/269 - Train Accuracy: 0.8581, Validation Accuracy: 0.8590, Loss: 0.2558
Epoch   6 Batch  187/269 - Train Accuracy: 0.8507, Validation Accuracy: 0.8593, Loss: 0.2472
Epoch   6 Batch  188/269 - Train Accuracy: 0.8715, Validation Accuracy

Epoch   6 Batch  267/269 - Train Accuracy: 0.8716, Validation Accuracy: 0.8581, Loss: 0.2463
Epoch   7 Batch    1/269 - Train Accuracy: 0.8513, Validation Accuracy: 0.8599, Loss: 0.2391
Epoch   7 Batch    2/269 - Train Accuracy: 0.8565, Validation Accuracy: 0.8652, Loss: 0.2411
Epoch   7 Batch    3/269 - Train Accuracy: 0.8669, Validation Accuracy: 0.8638, Loss: 0.2525
Epoch   7 Batch    4/269 - Train Accuracy: 0.8494, Validation Accuracy: 0.8588, Loss: 0.2455
Epoch   7 Batch    5/269 - Train Accuracy: 0.8512, Validation Accuracy: 0.8586, Loss: 0.2418
Epoch   7 Batch    6/269 - Train Accuracy: 0.8861, Validation Accuracy: 0.8611, Loss: 0.2249
Epoch   7 Batch    7/269 - Train Accuracy: 0.8654, Validation Accuracy: 0.8567, Loss: 0.2243
Epoch   7 Batch    8/269 - Train Accuracy: 0.8769, Validation Accuracy: 0.8623, Loss: 0.2462
Epoch   7 Batch    9/269 - Train Accuracy: 0.8591, Validation Accuracy: 0.8580, Loss: 0.2412
Epoch   7 Batch   10/269 - Train Accuracy: 0.8652, Validation Accuracy

Epoch   7 Batch   89/269 - Train Accuracy: 0.8853, Validation Accuracy: 0.8665, Loss: 0.2191
Epoch   7 Batch   90/269 - Train Accuracy: 0.8565, Validation Accuracy: 0.8578, Loss: 0.2326
Epoch   7 Batch   91/269 - Train Accuracy: 0.8788, Validation Accuracy: 0.8592, Loss: 0.2166
Epoch   7 Batch   92/269 - Train Accuracy: 0.8952, Validation Accuracy: 0.8657, Loss: 0.2162
Epoch   7 Batch   93/269 - Train Accuracy: 0.8770, Validation Accuracy: 0.8689, Loss: 0.2097
Epoch   7 Batch   94/269 - Train Accuracy: 0.8626, Validation Accuracy: 0.8685, Loss: 0.2349
Epoch   7 Batch   95/269 - Train Accuracy: 0.8701, Validation Accuracy: 0.8555, Loss: 0.2156
Epoch   7 Batch   96/269 - Train Accuracy: 0.8496, Validation Accuracy: 0.8479, Loss: 0.2206
Epoch   7 Batch   97/269 - Train Accuracy: 0.8749, Validation Accuracy: 0.8704, Loss: 0.2160
Epoch   7 Batch   98/269 - Train Accuracy: 0.8715, Validation Accuracy: 0.8701, Loss: 0.2202
Epoch   7 Batch   99/269 - Train Accuracy: 0.8622, Validation Accuracy

Epoch   7 Batch  178/269 - Train Accuracy: 0.8865, Validation Accuracy: 0.8734, Loss: 0.2008
Epoch   7 Batch  179/269 - Train Accuracy: 0.8745, Validation Accuracy: 0.8733, Loss: 0.2028
Epoch   7 Batch  180/269 - Train Accuracy: 0.8963, Validation Accuracy: 0.8728, Loss: 0.1931
Epoch   7 Batch  181/269 - Train Accuracy: 0.8565, Validation Accuracy: 0.8755, Loss: 0.2100
Epoch   7 Batch  182/269 - Train Accuracy: 0.8826, Validation Accuracy: 0.8788, Loss: 0.2023
Epoch   7 Batch  183/269 - Train Accuracy: 0.8941, Validation Accuracy: 0.8725, Loss: 0.1738
Epoch   7 Batch  184/269 - Train Accuracy: 0.8792, Validation Accuracy: 0.8731, Loss: 0.2044
Epoch   7 Batch  185/269 - Train Accuracy: 0.8881, Validation Accuracy: 0.8762, Loss: 0.2001
Epoch   7 Batch  186/269 - Train Accuracy: 0.8780, Validation Accuracy: 0.8770, Loss: 0.2026
Epoch   7 Batch  187/269 - Train Accuracy: 0.8771, Validation Accuracy: 0.8820, Loss: 0.1993
Epoch   7 Batch  188/269 - Train Accuracy: 0.8837, Validation Accuracy

Epoch   7 Batch  267/269 - Train Accuracy: 0.8893, Validation Accuracy: 0.8837, Loss: 0.1985
Epoch   8 Batch    1/269 - Train Accuracy: 0.8727, Validation Accuracy: 0.8876, Loss: 0.1900
Epoch   8 Batch    2/269 - Train Accuracy: 0.8836, Validation Accuracy: 0.8887, Loss: 0.2008
Epoch   8 Batch    3/269 - Train Accuracy: 0.8946, Validation Accuracy: 0.8940, Loss: 0.1957
Epoch   8 Batch    4/269 - Train Accuracy: 0.8735, Validation Accuracy: 0.8933, Loss: 0.1913
Epoch   8 Batch    5/269 - Train Accuracy: 0.8761, Validation Accuracy: 0.8885, Loss: 0.1925
Epoch   8 Batch    6/269 - Train Accuracy: 0.9023, Validation Accuracy: 0.8809, Loss: 0.1815
Epoch   8 Batch    7/269 - Train Accuracy: 0.8849, Validation Accuracy: 0.8853, Loss: 0.1892
Epoch   8 Batch    8/269 - Train Accuracy: 0.8933, Validation Accuracy: 0.8833, Loss: 0.1946
Epoch   8 Batch    9/269 - Train Accuracy: 0.8826, Validation Accuracy: 0.8870, Loss: 0.1922
Epoch   8 Batch   10/269 - Train Accuracy: 0.8915, Validation Accuracy

Epoch   8 Batch   89/269 - Train Accuracy: 0.8990, Validation Accuracy: 0.8873, Loss: 0.1742
Epoch   8 Batch   90/269 - Train Accuracy: 0.8710, Validation Accuracy: 0.8824, Loss: 0.1807
Epoch   8 Batch   91/269 - Train Accuracy: 0.8910, Validation Accuracy: 0.8872, Loss: 0.1693
Epoch   8 Batch   92/269 - Train Accuracy: 0.9083, Validation Accuracy: 0.8841, Loss: 0.1647
Epoch   8 Batch   93/269 - Train Accuracy: 0.8936, Validation Accuracy: 0.8774, Loss: 0.1623
Epoch   8 Batch   94/269 - Train Accuracy: 0.8838, Validation Accuracy: 0.8817, Loss: 0.1856
Epoch   8 Batch   95/269 - Train Accuracy: 0.8821, Validation Accuracy: 0.8924, Loss: 0.1682
Epoch   8 Batch   96/269 - Train Accuracy: 0.8708, Validation Accuracy: 0.8857, Loss: 0.1786
Epoch   8 Batch   97/269 - Train Accuracy: 0.8960, Validation Accuracy: 0.8804, Loss: 0.1710
Epoch   8 Batch   98/269 - Train Accuracy: 0.8878, Validation Accuracy: 0.8897, Loss: 0.1757
Epoch   8 Batch   99/269 - Train Accuracy: 0.8850, Validation Accuracy

Epoch   8 Batch  178/269 - Train Accuracy: 0.9027, Validation Accuracy: 0.8944, Loss: 0.1677
Epoch   8 Batch  179/269 - Train Accuracy: 0.8851, Validation Accuracy: 0.8888, Loss: 0.1556
Epoch   8 Batch  180/269 - Train Accuracy: 0.9077, Validation Accuracy: 0.8960, Loss: 0.1484
Epoch   8 Batch  181/269 - Train Accuracy: 0.8790, Validation Accuracy: 0.8912, Loss: 0.1685
Epoch   8 Batch  182/269 - Train Accuracy: 0.9073, Validation Accuracy: 0.8909, Loss: 0.1585
Epoch   8 Batch  183/269 - Train Accuracy: 0.9042, Validation Accuracy: 0.8903, Loss: 0.1404
Epoch   8 Batch  184/269 - Train Accuracy: 0.9052, Validation Accuracy: 0.8906, Loss: 0.1632
Epoch   8 Batch  185/269 - Train Accuracy: 0.9000, Validation Accuracy: 0.8965, Loss: 0.1591
Epoch   8 Batch  186/269 - Train Accuracy: 0.8896, Validation Accuracy: 0.8934, Loss: 0.1577
Epoch   8 Batch  187/269 - Train Accuracy: 0.8935, Validation Accuracy: 0.8929, Loss: 0.1554
Epoch   8 Batch  188/269 - Train Accuracy: 0.9083, Validation Accuracy

Epoch   8 Batch  267/269 - Train Accuracy: 0.8969, Validation Accuracy: 0.8968, Loss: 0.1647
Epoch   9 Batch    1/269 - Train Accuracy: 0.8939, Validation Accuracy: 0.9006, Loss: 0.1459
Epoch   9 Batch    2/269 - Train Accuracy: 0.8898, Validation Accuracy: 0.8960, Loss: 0.1518
Epoch   9 Batch    3/269 - Train Accuracy: 0.9032, Validation Accuracy: 0.8963, Loss: 0.1484
Epoch   9 Batch    4/269 - Train Accuracy: 0.8828, Validation Accuracy: 0.8999, Loss: 0.1549
Epoch   9 Batch    5/269 - Train Accuracy: 0.8881, Validation Accuracy: 0.9062, Loss: 0.1509
Epoch   9 Batch    6/269 - Train Accuracy: 0.9015, Validation Accuracy: 0.8971, Loss: 0.1386
Epoch   9 Batch    7/269 - Train Accuracy: 0.8890, Validation Accuracy: 0.8965, Loss: 0.1402
Epoch   9 Batch    8/269 - Train Accuracy: 0.9062, Validation Accuracy: 0.9041, Loss: 0.1511
Epoch   9 Batch    9/269 - Train Accuracy: 0.9011, Validation Accuracy: 0.8992, Loss: 0.1522
Epoch   9 Batch   10/269 - Train Accuracy: 0.9120, Validation Accuracy

Epoch   9 Batch   89/269 - Train Accuracy: 0.9208, Validation Accuracy: 0.9023, Loss: 0.1387
Epoch   9 Batch   90/269 - Train Accuracy: 0.8962, Validation Accuracy: 0.9069, Loss: 0.1454
Epoch   9 Batch   91/269 - Train Accuracy: 0.9010, Validation Accuracy: 0.9022, Loss: 0.1313
Epoch   9 Batch   92/269 - Train Accuracy: 0.9299, Validation Accuracy: 0.8978, Loss: 0.1306
Epoch   9 Batch   93/269 - Train Accuracy: 0.9162, Validation Accuracy: 0.9032, Loss: 0.1320
Epoch   9 Batch   94/269 - Train Accuracy: 0.8999, Validation Accuracy: 0.9013, Loss: 0.1455
Epoch   9 Batch   95/269 - Train Accuracy: 0.9024, Validation Accuracy: 0.9052, Loss: 0.1373
Epoch   9 Batch   96/269 - Train Accuracy: 0.8868, Validation Accuracy: 0.9034, Loss: 0.1345
Epoch   9 Batch   97/269 - Train Accuracy: 0.9039, Validation Accuracy: 0.8990, Loss: 0.1340
Epoch   9 Batch   98/269 - Train Accuracy: 0.9059, Validation Accuracy: 0.8974, Loss: 0.1371
Epoch   9 Batch   99/269 - Train Accuracy: 0.9033, Validation Accuracy

Epoch   9 Batch  178/269 - Train Accuracy: 0.9167, Validation Accuracy: 0.9079, Loss: 0.1244
Epoch   9 Batch  179/269 - Train Accuracy: 0.8938, Validation Accuracy: 0.9077, Loss: 0.1274
Epoch   9 Batch  180/269 - Train Accuracy: 0.9221, Validation Accuracy: 0.9058, Loss: 0.1163
Epoch   9 Batch  181/269 - Train Accuracy: 0.8931, Validation Accuracy: 0.9007, Loss: 0.1399
Epoch   9 Batch  182/269 - Train Accuracy: 0.8999, Validation Accuracy: 0.9048, Loss: 0.1321
Epoch   9 Batch  183/269 - Train Accuracy: 0.9196, Validation Accuracy: 0.9124, Loss: 0.1067
Epoch   9 Batch  184/269 - Train Accuracy: 0.9179, Validation Accuracy: 0.9111, Loss: 0.1320
Epoch   9 Batch  185/269 - Train Accuracy: 0.9147, Validation Accuracy: 0.9114, Loss: 0.1260
Epoch   9 Batch  186/269 - Train Accuracy: 0.9111, Validation Accuracy: 0.9060, Loss: 0.1273
Epoch   9 Batch  187/269 - Train Accuracy: 0.9043, Validation Accuracy: 0.9079, Loss: 0.1261
Epoch   9 Batch  188/269 - Train Accuracy: 0.9148, Validation Accuracy

Epoch   9 Batch  267/269 - Train Accuracy: 0.9262, Validation Accuracy: 0.9149, Loss: 0.1324
Model Trained and Saved


### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [20]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    word_ids = []
    for word in sentence.lower().split():
        if word in vocab_to_int:
            word_ids.append(vocab_to_int[word])
        else:
            word_ids.append(vocab_to_int['<UNK>'])
    return word_ids


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [21]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [228, 23, 55, 107, 34, 140, 162]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [275, 16, 32, 223, 296, 57, 108, 324, 1]
  French Words: il a vu à l' voiture mai . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.